### PR TITLE
Add decoders module

### PR DESCRIPTION
**Pull request by thefliik**
This is a PR for #491.

This module facilitates decoding an `unknown` input and returning a properly typed response (on success), or returning any errors. The API is generic, composable, and promise/async friendly. Obvious use cases would be anytime you are receiving unknown/unverified data, such as CLI arguments or web request params, but also includes things like reading data from a file. More broadly, it aims to provide a unified interface for decoders of any type.

The [readme](https://github.com/thefliik/deno_std/tree/decoders/decoders#decoders) provides an overview of the module's proposed API and functionality. The actual [`Decoder` interface](https://github.com/thefliik/deno_std/tree/decoders/decoders#interfaces) is very simple. Most of this module's power comes from an assortment of decoder-creating functions which it exports (e.g. `isString()`, `isObject()`).

- A special focus has been given on making sure that returned error messages are useful and easily customizable. See [working with errors](https://github.com/thefliik/deno_std/tree/decoders/decoders#working-with-errors) in the readme.
- When doing async decoding, decoders provide the option of decoding values in parallel if appropriate.

### Somewhat complex example

[_Also see `parseDate()` from deno_std's datetime module implemented as a decoder, below._](https://github.com/denoland/deno_std/pull/522#issuecomment-505204206)

```ts
const shortStringDecoder = isChainOf([
  isString(),
  isMatchForPredicate(value => value.length !== 0 && value.length < 50),
]);

const decoder = isObject({
  userId: isString(),
  type: isExactly("address"),
  payload: isObject({
    latLon: isTuple([ isNumber(), isNumber() ]),
    locationName: isOptional(shortStringDecoder),
    street: shortStringDecoder,
    // validates that input is a string in YYYY-MM-DD format and, if valid, maps to a `Date` object.
    updatedOn: isMatch(/^\d{4}[\/\-](0?[1-9]|1[012])[\/\-](0?[1-9]|[12][0-9]|3[01])$/)
      .map(value => new Date(value))
  })
}, { allErrors: true });

const result = decoder.decode({ 
  userId: '123',
  type: 'address',
  payload: {
    latLon: [ 30.4, 120.3 ],
    street: '123 awesome street',
    city: 'amazing city', // <-- excess property
    updatedOn: '2019-06-10',
  }
})

result instanceof DecoderSuccess === true;

// Type of result.value
// {
//   userId: string,
//   type: 'address',
//   payload: {
//     latLong: [number, number],
//     locationName: string | undefined,
//     street: string;
//     updatedOn: Date
//   }
// }
```

## Reasons to include a module like this in deno_std

1. Dealing with an `unknown` input is common across application contexts and especially in a scripting environment.
   - Providing a convenient method for decoding `unknown` data and *typing the response* is broadly useful.
   - Creating error messages for `unknown` or malformed data can be time consuming. Providing a module with out-of-the-box error messages, as well as robust error message customizability is broadly useful.
2. Decoding `unknown` inputs touches on security, and having that module maintained in deno_std would be a plus.
3. One of the strengths of a good decoder API is its composibility. Marking a decoder API as "official" in deno_std could (hopefully) facilitate interoperability between libraries and encourage composibility. Examples:
   1. A well made CSV parsing decoder could be used stand-alone to decode CSV data or it could be composed with other decoders to validate incoming web params and parse the CSV data inside.
   2. A custom date library, e.g. moment, could provide a decoder which validates time-related inputs and returns appropriate date objects. 
   3. You could combine the hypothetical CSV parsing decoder and the hypothetical date library decoder, along with the decoders included in this module, to parse a csv file and transform the contents of cells containing dates into the appropriate Date objects.

### CODE REVIEW

**Comment by thefliik** [comment]
@j-f1 not sure how interested you are in issue #491, but here's a ping.

**Comment by thefliik** [comment]
Quick and dirty example of implementing `parseDate`  from the deno_std datetime module as a decoder:

```ts
function isDate(format: 'yyyy-mm-dd' | 'dd-mm-yyyy') {
  if (format === "yyyy-mm-dd") {
    return isRegex(
             /^\d{4}[\/\-](0?[1-9]|1[012])[\/\-](0?[1-9]|[12][0-9]|3[01])$/,
             { msg: "invalid date - must be in \"YYYY-MM-DD\" format" }
           )
             .map(value => new Date(value))
  }
  else if (format === "dd-mm-yyyy") {
    return isRegex(
             /^(0?[1-9]|[12][0-9]|3[01])[\/\-](0?[1-9]|1[012])[\/\-]\d{4}$/,
             { msg: "invalid date - must be in \"DD-MM-YYYY\" format" }
           )
             .map(value => new Date(value.split('').reverse().join(''))
  }

  throw new Error("Invalid format argument)"
}
```
simple usage
```ts
const parseDate = assert(isDate("yyyy-mm-dd"));

const date = parseDate("2019-06-10"); // returns Date object
const date = parseDate("10-06-2019"); // throws DecoderError "invalid date - must be in \"YYYY-MM-DD\" format"
```
using it combined with other decoders
```ts
const decoder = isObject({
  payload: {
    type: isString(),
    value: isAny(),
  },
  updatedOn: isDate("yyyy-mm-dd")
})
```

**Comment by j-f1** [comment]
Why did you decide on the `is` prefix instead of having no prefix?

File: type_decoders/README.md [hunk 1]
```diff
@@ -0,0 +1,787 @@
...
+interface INullDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isNull(options?: INullDecoderOptions): Decoder<null, unknown>;
+```
+
+`isNull()` can be used to verify that an unknown value is `null`. This is a convenience function for `isExactly(null)`.
+
+### isRegex()
```
Comment (j-f1) [inline type_decoders/README.md]: Should this be called `isMatch` or similar, since it doesn’t check the input to see if it’s a regular expression, but rather matches the passed regular expression against the input?

File: type_decoders/README.md [hunk 2]
```diff
@@ -0,0 +1,787 @@
...
+
+```ts
+interface IRegexDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isRegex(regex: RegExp, options?: IRegexDecoderOptions): Decoder<string, unknown>;
+```
+
+`isRegex()` can be used to verify that an unknown value is `string` which conforms to the given `RegExp`.
```
Comment (j-f1) [inline type_decoders/README.md]: I’m not exactly sure how much you’re allowing matchers to overlap, but would it make sense to re-type this as taking in a string and outputting a string, leaving the actual check to see if the input is a string to `isString`? Or would that be too much DRY?

File: type_decoders/README.md [hunk 3]
```diff
@@ -0,0 +1,787 @@
...
+Example:
+
+```ts
+const validator = assert(isObject({a: isString() b: isOptional(isNever()) }));
+
+validator({ a: 'one' }) // returns { a: 'one' }
+validator({ a: 'one', b: 'two' }) // throws DecoderError
+```
+
+### isExactly()
```
Comment (j-f1) [inline type_decoders/README.md]: Maybe call this `is`?

File: type_decoders/README.md [hunk 4]
```diff
@@ -0,0 +1,787 @@
...
+interface IExactlyDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isExactly<T>(value: T, options?: IExactlyDecoderOptions): Decoder<T, unknown>;
+```
+
+`isExactly()` accepts a `value: T` argument and can be used to verify that an unknown input is `=== value`.
+
+### isConstant()
```
Comment (j-f1) [inline type_decoders/README.md]: Maybe `isOverriddenToBe` or something more concise with the same meaning?

File: type_decoders/README.md [hunk 5]
```diff
@@ -0,0 +1,787 @@
...
+interface IInstanceOfDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isInstanceOf<T extends new (...args: any) => any>(clazz: T, options?: IInstanceOfDecoderOptions): Decoder<InstanceType<T>, unknown>;
+```
+
+`isInstanceOf()` accepts a javascript constructor argument and creates a decoder which verifies that its input is `instanceof clazz`.
+
+### isCheckedWith()
```
Comment (j-f1) [inline type_decoders/README.md]: `isMatchForPredicate`?

If we got rid of the `is` prefix, this could be `matchesPredicate`.

File: type_decoders/README.md [hunk 6]
```diff
@@ -0,0 +1,787 @@
...
+```ts
+interface IAnyOfDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isAnyOf<T extends Decoder<unknown>>(decoders: T[], options?: IAnyOfDecoderOptions): Decoder<DecoderReturnType<T>>;
+function isAnyOf<T extends Decoder<unknown> | PromiseDecoder<unknown>>(decoders: T[], options?: IAnyOfDecoderOptions): PromiseDecoder<DecoderReturnType<T>>;
+```
+
+`isAnyOf()` accepts an array of decoders and attempts to decode a provided value using each of them, in order, returning the first successful result or `DecoderError` if all fail.
```
Comment (j-f1) [inline type_decoders/README.md]: Which `DecoderError` would be returned here? Would it create its own or somehow combine the errors of the passed decoders?

File: type_decoders/README.md [hunk 7]
```diff
@@ -0,0 +1,787 @@
...
+  keyMap?: { [P in keyof T]?: string | number };
+}
+
+function isObject<T>(decoderObject: { [P in keyof T]: Decoder<T[P]> }, options?: IObjectDecoderOptions<T>): Decoder<T> ;
+function isObject<T>(decoderObject: { [P in keyof T]: Decoder<T[P]> | PromiseDecoder<T[P]> }, options?: IObjectDecoderOptions<T>): PromiseDecoder<T> ;
+```
+
+`isObject()` accepts a `key: Decoder` argument object and returns a new decoder that will verify that an input is a non-null object, that the input has the keys specified in the argument object, and that the key values pass the relevant decoder provided by the argument object. On `DecoderSuccess`, the a new object is returned and the key-values of that object are provided by output from the argument decoders.
+
+### isExactObject()
```
Comment (j-f1) [inline type_decoders/README.md]: Would it make sense to make this an additional option to `isObject`, i.e. `isObject({ ... }, { exact: true })`?

File: type_decoders/README.md [hunk 8]
```diff
@@ -0,0 +1,787 @@
...
+
+### isDictionary()
+
+```ts
+interface IDictionaryDecoderOptions<V> {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isDictionary<R, V = unknown>(decoder: Decoder<R, V>, options?: IDictionaryDecoderOptions<V>): Decoder<R[], V>;
+function isDictionary<R, V = unknown>(decoder: PromiseDecoder<R, V>, options?: IDictionaryDecoderOptions<V>): PromiseDecoder<R[], V>;
```
Comment (j-f1) [inline type_decoders/README.md]: It looks like the return types here have a typo — should they be `{ [key: string]: R }` instead of `R[]`?

File: type_decoders/README.md [hunk 9]
```diff
@@ -0,0 +1,787 @@
...
+```
+
+`isDictionary()` receives a decoder argument and uses that decoder to process all values (regardless of key) of an input object.
+
+### isArray()
```
Comment (j-f1) [inline type_decoders/README.md]: `isArrayOf`?

File: type_decoders/README.md [hunk 10]
```diff
@@ -0,0 +1,787 @@
...
+}
+
+function isArray<R = unknown, V = unknown>(options?: IArrayDecoderOptions<V>): Decoder<R[], V>;
+function isArray<R, V = unknown>(decoder: Decoder<R, V>, options?: IArrayDecoderOptions<V>): Decoder<R[], V>;
+function isArray<R, V = unknown>(decoder: PromiseDecoder<R, V>, options?: IArrayDecoderOptions<V>): PromiseDecoder<R[], V>;
+```
+
+`isArray()` can be used to make sure an input is an array. If an optional decoder argument is provided, that decoder will be used to process all of the input's elements.
+
+### isTuple()
```
Comment (j-f1) [inline type_decoders/README.md]: `isTupleOf`?

File: type_decoders/README.md [hunk 11]
```diff
@@ -0,0 +1,787 @@
...
+  keyMap?: { [P in keyof T]?: string | number };
+}
+
+function isExactObject<T>(decoderObject: { [P in keyof T]: Decoder<T[P]> }, options?: IExactObjectDecoderOptions<T>): Decoder<T> ;
+function isExactObject<T>(decoderObject: { [P in keyof T]: Decoder<T[P]> | PromiseDecoder<T[P]> }, options?: IExactObjectDecoderOptions<T>): PromiseDecoder<T> ;
+```
+
+`isExactObject()` is the same as `isObject()`, except the input object cannot have any excess properties.
+
+### isDictionary()
```
Comment (j-f1) [inline type_decoders/README.md]: `isDictionaryOf`?

Also, would it make sense to have an optional second decoder parameter to check the keys?

```js
isDictionaryOf(isNumber(), isRegex(/^[a-z]+$/) // enforce camelCase keys
```

File: type_decoders/README.md [hunk 12]
```diff
@@ -0,0 +1,787 @@
...
+```
+
+`isTuple()` receives an array of decoders and creates a decoder which can be used to verify that an input is:
+
+1. An array of the same length as the decoder argument array.
+2. The first decoder argument will be used the process the first element of an input array.
+3. The second decoder argument will be used the process the second element of an input array.
+4. etc...
+
+### isLazy()
```
Comment (j-f1) [inline type_decoders/README.md]: When would this be used? Would it be when the `Decoder` tree is expensive to generate?

**Comment by thefliik** [comment]
> Why did you decide on the `is` prefix instead of having no prefix?

Good question. Unprefixed, a few of the function names would potentially conflict with javascript keywords (e.g. `isNull() => null()` or `isUndefined() => undefined()`) so in at least a few circumstances some sort of prefix/name change would be necessary.

One solution would be to only change the names of functions that needed it (e.g. `null_()`). To the extent possible though, I would prefer an API which is consistent and predictable. This leads to some sort of naming scheme.

I originally played around with simply adding the suffix `Decoder` to the end of functions (e.g. `stringDecoder()` and `objectDecoder()`). I found this created a lot of noise in the API though, and made it harder for me to visually parse the information when dealing with a large decoder:

Example:

```ts
objectDecoder({
  id: stringDecoder(),
  type: exactlyDecoder('address'),
  payload: objectDecoder({
    value1: numberDecoder()
    value2: numberDecoder()
  })
})
```

I also tried the shorter suffix `Dec` as in:

```ts
objectDec({
  id: stringDec(),
  type: exactlyDec('address'),
  payload: objectDec({
    value1: numberDec()
    value2: numberDec()
  })
})
```

I definitely like this more, but ultimately I felt the prefix `is` was more visually appealing and it seems to more clearly describe what the decoder is actually doing i.e. `isObject({})` is creating a decoder to assert that something is an object.

```ts
isObject({
  id: isString(),
  type: isExactly('address'),
  payload: isObject({
    value1: isNumber()
    value2: isNumber()
  })
})
```

Part of the reason why I think the `is` prefix is easy to read, is that the prefix is short, all lower case, and makes it easy to pick out the real descriptor in the name: the noun part of the name which begins with a capital letter. i.e. in `isString()`, it's easy to visually pick out the `String()` part of the name.

The main con of the `is` prefix (that I've thought of) is that the naming format `isSomething()` is sometimes associated with predicate functions. Still, this con doesn't seem to outweigh the upsides of the prefix, and decoders do seem to share a certain thematic similarity with a predicate function (both are asserting something).

I'm definitely open to feedback / suggestions though.

File: type_decoders/README.md (hunk 1 above)
Comment (thefliik) [reply type_decoders/README.md]: @j-f1 good suggestion

File: type_decoders/README.md (hunk 2 above)
Comment (thefliik) [reply type_decoders/README.md]: It's a good question. This *might* (not sure) be the only decoder which overlaps in this way and I did think about it a bit. Ultimately, I feel like that would be too dry. For 1) I think most uses of this decoder will be to simply check that something is a string of a certain format. Requiring someone to manually compose isChainOf([isString(), isRegex()]) all the time would be annoying. Possibly more importantly, currently all the exported decoders in the module expect `unknown` inputs. At the moment, I think there is value in maintaining that consistency, even if it results in the occasional, unnecessary, `isString()` check. I.e. predictability as a feature.

**Comment by j-f1** [comment]
One option would be to get rid of the `isUndefined`/`isNull` decoders and instead use `is(undefined)`/`is(null)`, especially since they’d be much more frequently used<sup>[*citation needed*]</sup> in the form of `isOptional`/`isNullable`. This would allow the use of unprefixed function names.

File: type_decoders/README.md (hunk 6 above)
Comment (thefliik) [reply type_decoders/README.md]: `isAnyOf` is unique in that it doesn't return the nested `DecoderErrors`, and instead returns it's own error. By default, the error message is `(child error messages[]).join(' OR ')`. A current limitation of this module's API is that only a single `DecoderError` (the first one) is returned (i.e. you can't collect all the errors associated with an input). Because of this, it isn't currently possible to return multiple nested errors with a isAnyOf` DecoderError. I'm pretty convinced that the final API should support multiple nested errors though. Though this shouldn't be a large change.

File: type_decoders/README.md (hunk 8 above)
Comment (thefliik) [reply type_decoders/README.md]: Good catch

File: type_decoders/README.md (hunk 7 above)
Comment (thefliik) [reply type_decoders/README.md]: Yup, another good suggestion

File: type_decoders/README.md (hunk 9 above)
Comment (thefliik) [reply type_decoders/README.md]: `isArray()` actually supports being called without arguments (to test if something is an array but ignore what is inside). More broadly though, I'm unconvinced that `isObject`, `isArray`, `isTuple`, `isDictionary`, etc need an "Of" suffix for clarity?

The only reason why I added "Of" to `isAnyOf([])` and `isChainOf([])` was because they felt a little confusing without it, but that doesn't seem to be the same for `isObject` etc. For one, `isArray`, `isObject`, etc are all concrete *things* whereas `isAnyOf()` / `isChainOf()` are more abstract concepts. Definitely open to feedback here though.

File: type_decoders/README.md (hunk 11 above)
Comment (thefliik) [reply type_decoders/README.md]: Ooo. Good suggestion re: `optional second decoder parameter`.

Regarding `isDictionaryOf`, see my reply to your `isArrayOf` comment.

File: type_decoders/README.md (hunk 12 above)
Comment (thefliik) [reply type_decoders/README.md]: Good question. I should make the description clearer by providing an example. It's for a recursive data structure. For example:

```ts
const treeDecoder = isObject({
  value: isString(),
  children: isArray(isLazy(() => treeDecoder))
})
```

This object would have the interface 

```ts
interface ITreeDecoder {
  value: string;
  children: ITreeDecoder[];
}
```

Comment (thefliik) [reply type_decoders/README.md]: It could be named `isRecursive`. I chose the name because its lazily evaluating the decoder, but on reflection `isRecursive` probably makes more sense.

File: type_decoders/README.md (hunk 6 above)
Comment (j-f1) [reply type_decoders/README.md]: It’s also a little difficult to convey since you’d normally expect at least some of the decoders to fail even if the `anyOf` succeeds.

File: type_decoders/README.md (hunk 9 above)
Comment (j-f1) [reply type_decoders/README.md]: IMO `isArray` is more abstract, whereas `isTuple` is the more concrete version.

**Comment by thefliik** [comment]
> One option would be to get rid of the `isUndefined`/`isNull` decoders and instead use `is(undefined)`/`is(null)`

That's a good workaround I hadn't thought of. Lemme see how removing the prefixes looks and gelling with it a bit. Originally, I didn't love the `is` prefix, but now I find I've gotten used to it.

One additional thing to think about: if this module (or a module like it) is added to deno_std, I hope one of the goals will be to passively promote a standard `Decoder` API, so that other modules/libraries can add decoders that are composable with this module and each other. For example, a `uuid` module could provide a uuid decoder. A date library could provide a decoder for its custom date type, etc. Providing some kind of naming strategy for this library's decoders could establish a useful decoder naming convention for deno. For example, if a `uuid` module exported an `isUuid()` module you might (hypothetically) have a clue that it was a decoder.

Maybe this is overthinking things though. I imagine the natural state of things would be for a `uuid` module to export a decoder named `uuidDecoder()`, which is certainly clear.

**Comment by j-f1** [comment]
Good point about the custom decoders 👍 Maybe the `is` prefix is the way to go, after all.

**Comment by thefliik** [comment]
@ry no rush, but this PR is ready for review. I notice the CI tests are failing but the failing tests are all in other `deno_std` modules. All of this module's tests are passing on my computer, and I did not modify any files outside of the new `type_decoders` directory (other than to wire up the tests in `./test.ts`). I did rebase off of master not too long ago which may have been a mistake. Anyway, any feedback is welcomed.

- FYI: the `any` type is intentionally used in various places in the module where someone might have expected to see `unknown`. In both cases, the type is representing an "unknown" value. The reason arises when composing decoders together: if a function expects a decoder which accepts an input string and returns a string, it should be able to accept a decoder which accepts an unknown input and returns a string. When "unknown input" is typed as `unknown`, attempting to assign `Decoder<string, unknown>` to `Decoder<string, string>` will result in a type error. Assigning `Decoder<string, any>` to `Decoder<string, string>` does not raise a type error.

CC: @j-f1 any additional feedback is welcomed. A lot of your feedback I incorporated. Some of it, like dropping the `is` prefix naming convention, I decided to wait on to see what other people think. Also, this module now supports returning all errors, as well as doing async decoding in parallel.

File: type_decoders/README.md (hunk 12 above)
Comment (thefliik) [reply type_decoders/README.md]: I ended up changing the name back to `isLazy()`. This is because while I imagine the primary use case will be decoding recursive data structures, what the decoder is actually *doing* is lazily evaluating a function each time `decode()` is called to retrieve the decoder it should use. I imagine `isLazy()` has other use cases beyond simply recursive data structures.

I added some more detail in the API explanation to describe the must common use case.

**Comment by thefliik** [comment]
@bios21, @zekth, @bartlomieju, given your involvement in the YAML decoder PR, feel free to contribute any feedback you have to this PR (though no pressure!).

### tldr

This PR is attempting to do 2 things

1. Create a [standard interface for deno decoders](https://github.com/thefliik/deno_std/tree/decoders/decoders#interfaces) (through two base classes). Here the term "decoding" refers to both parsing and transforming input (e.g. CSV string to `Array` or date string to `Date`), as well as validating input to make sure it has a certain form.
2. Add an assortment of useful decoders which implement this new API. The decoders I've included could largely be called JSON decoders, though I've chosen to simply call them "decoders" since they are working with standard javascript objects.

If you're not sure what the benefit of creating a standard decoder interface would be, here are some points:

1.  Decoders are common. Dealing with an `unknown` input is common across application contexts and especially in a scripting environment.
2. It is common to be dealing with mixed input.
   - CSV which contains `Date` cells
   - JSON params which contain CSV data
   - YAML data which contains UUIDs.
3. It can be tricky to get good error messages from invalid data. Where "good" can mean both human readable as well as machine-readable.
4. In typescript, I want my decoded values to be properly typed.

The right decoder standard can help address these issues, as well as create conventions which other modules/libraries can follow.

Anyway, happy to expand on any of this. I'll also say that this PR is just one proposal for how to create a standard decoder API. I'm definitely open to other ideas.

Final thought: this PR is pretty large but the actual decoder interface / API it establishes is small. You can get a good overview of everything by reading the [introductory comment above](https://github.com/denoland/deno_std/pull/522#issue-291289571), as well as the readme.
